### ISSUE TITLE
Xiangqi chasing rules

### ISSUE THREAD

**Issue by ianfab** [ianfab/Fairy-Stockfish#55]
Support more chasing rules besides perpetual check. The exact set of chasing rules to be supported is an open question.

References:
http://wxf.ca/wxf/doc/book/WXF_rules_Eng.pdf
http://www.xqinenglish.com/images/Downloads/ChineseChessRules.pdf
http://www.asianxiangqi.org/English/AXF_rules_Eng.pdf
http://www.talkchess.com/forum3/viewtopic.php?t=35403

An eventual implementation could potentially also be used for Janggi #142.

**Comment by gaintpd** [comment]
The AXF rule is relatively easy compared to Chinese Xiangqi Federation Rule（in mainland china） though it is still very difficult to understand

**Comment by Marken-Foo** [comment]
My summary of the AXF rules (extrapolated from all the examples, diagram numbers provided with #D)

**Executive summary**:
Perpetual check loses (unless both sides check perpetually, #D4)
Perpetual chase usually* loses (see clarifications)
Perpetual check+chase is acceptable. (e.g. #D12)

**What is a chasing move?**
An attack on an unprotected piece, *or* an attack on a chariot by a weaker piece (#D25).
Pinned pieces do not protect other pieces (#D83-85).
Pinned pieces do not attack other pieces (*even if they can legally capture said piece!*) (#D63, #D39-41)

**Perpetual chase**
Multiple pieces involved:
- 1v2, 2v2, 1v3 etc allowed: Chasing 2 different pieces (one on each move) is allowed. (#D80)
- 2v1, 3v1 etc illegal. (what I call "bullying") -- multiple pieces taking turns to chase the same piece every move.
1v1 chasing is illegal, unless:
- Chasing piece is a king or pawn, then allowed. (#D67-70)
    - But bullying is still illegal, overrides this (#D71-73)
- Same piece type chasing same piece type is allowed (#D51-53 if both unprotected; #D101-103 if only one unprotected)
    - But bullying is still illegal, overrides this (#D92-93)

**Caveats**
Chasing suicidal pieces is not really chasing (additional position #D2)
?????? (additional position #D3)
I also remember talking to the person who wrote the rules and attempting to point out some uncovered/contradictory cases, but lost my notes on those. These rules should cover most situations though.

Rules used: [AXF_rules_Eng.pdf](https://github.com/ianfab/Fairy-Stockfish/files/4207842/AXF_rules_Eng.pdf)

**Comment by HGMuller** [comment]
This issue seems to be closed, but is still advertized as 'help wanted'. So the help might be too late. Let me nevertheless comment on your summary:

I assume that with "even if it can legally captured said piece" you mean "even if it can legally capture its pinner".  This has nothing to do with it, however. The issue is whether the attack is new or not; attacks that existed before never count as chases, even when they temporarily were not legal because a check has to be resolved first. It doesn't matter whether you are pinned, but whether the capture is legal, and pinned pieces can have legal captures. It is just that illegal moves should be considered non-existent (except when they are illegal for no other reason than that they do not resolve an existing check).

When you say "bullying is still illegal" it also seems to miss the point. It is perfectly allowed to bully a piece (say a Cannon) with Pawns. Say black has a Cannon on c9, and white Pawns on b8 and d8. Then 1. bc8 Ce9 2. de8 Cb9 3. cb8 Cd9 4. ed8 Cc9 {repeat} 'bullies' the Cannon, but would force a draw if the Cannon for some reason was trapped in the range b9-e9. Because the attacks were made by Pawns, and thus should be ignored. In the examples you quote red is at fault, because even when the attacks by the King are ignored, this King discovers attacks by other pieces. And those attacks do count.

I would summarize the rules as follows:

- On a perpetual, if one side violates the rules _more severly_ than the other, he loses (and otherwise it is draw).
- Chasing a royal (i.e. checking) is a greater violation than chasing any other piece
- Chasing an _unpassed_ Pawn is no violation at all
- A chase is making a _new_ attack on the _same_ defenseless piece each and every move of the repeat cycle
- Only attacks that can _legally_ capture are taken into account
- Attacks are not new when they existed pseudo-legally before, but were illegal because an existing check had to be resolved
- Pieces are not defenseless when they can legally capture their attacker
- Rooks are defenseless against attacks by a Horse or Cannon
- In other cases a piece is defenseless if no legal recapture (i.e. on the same square) is possible after the attack captures it
- attacks by a King or a Pawn (but not by any pieces their moves discover!) should be ignored

Logically one would expect that attacks by A or E on a protected R are also chases. The rules do not mention this, though. This could be for 'never-happens' reasons. Although in theory a Rook can be chased by an Elephant through discovered attacks.

**Comment by gbtami** [comment]
@HGMuller this issue is definitely open (green!). Only #179 is closed (as duplicate). :)

**Comment by HGMuller** [comment]
I should also have pointed out that 'attacks' should be interpreted as (attacker, victim) pairs rather than as (fromSquare, toSquare) pairs. So an attack by the same piece in another location does not constitute a new attack.

**Comment by noobpwnftw** [comment]
> chessdb.cn adjudicates discovered attack of E against protected R to lose, though.

Can you provide an example of such position?

**Comment by cloudfish** [comment]
> > chessdb.cn adjudicates discovered attack of E against protected R to lose, though.
> 
> Can you provide an example of such position?

Hello! Are you the webmaster of chessdb.cn? Thank you for making this fantastic website!

A minimal example would consist of only KCB vs. KRR:

4k4/9/9/9/9/r1r6/3C5/4B4/9/5K3 w moves d3f3 c4g4 f3d3 g4c4 d3f3

Of course the moves here are ridiculous and nowhere near the principle variation. So I spent a little time making the following example, where red can indeed make his situation better if he is allowed to perpetual chase the protected black rook:

3ak4/4a4/9/9/2b6/nnrp2B2/1C1p5/B3N3C/3pA4/pp1A1K3 w moves b3b1 c4c0 b1b3 c0c4 b3b1

Both chessdb and Tiantian Xiangqi adjudicates to black in this situation. I think the AXF rules are to be taken literally and would consider this a bug.

**Comment by noobpwnftw** [comment]
Please refer to "http://www.asianxiangqi.org/English/AXF_rules_Eng.pdf", page 25, section 26 for the description of this particular rule. As I understand it, discovered attacks of cannons and knights do not care about whether the moving piece is a blocker or not, as long as the outcome results in a perpetual chase on a rook, it is a perpetual chase in violation of rules if no other situations(pinned/opponent chases/checks) apply.

**Comment by cloudfish** [comment]
The examples I provided are not discovered attacks of cannons; they are discovered attacks of elephants. In fact the term "discovered attack" is not relevant here; programmatically speaking (see also HGMuller's comment above), you only need to consider the new (attacker, victim) pair in the resulting position, which is (red elephant, black rook) in my examples. The mover's type or the blocker's type are not relevant. Per rule 3.(6) on page 4,

"Perpetual chase on a piece with real root will be declared a draw. It is a draw if the protector is a fake root (Translation mistake? Should be 'a loss'). However, perpetual chase on a chariot by a horse or a cannon is prohibited."

if the victim has a real root, it's a perpetual chase only if the (attacker, victim) pair is (horse/cannon, rook).

**Comment by gbtami** [comment]
quanghuy2309 posted this link on Fairy Discord, it may be useful here.
https://sci-hub.se/10.3233/icg-180054

**Comment by xiangqishare** [comment]
The two strongest engines in Xiangqi(cyclone/bugchess) disagree on chase rules

**Comment by xiangqishare** [comment]
> The two strongest engines in Xiangqi(cyclone/bugchess) disagree on chase rules Can you give an example where they disagree?

A few minimal examples
Chessdb＝Bugchess→draw. Cyclone→red violation:
4k4/4a4/4bn3/9/5R3/5C3/9/9/9/4K4 w moves f4e4 e9f9 e4f4 f9e9
1RC1k4/9/4n4/9/9/9/9/4B4/9/4K4 w moves b9b7 e7f9 b7b9 f9e7

chessdb＝Cy→draw. Bugchess→red violation:
4k4/9/4n4/9/3c5/9/4N4/9/9/3K5 w moves d0e0 e7c8 e0d0 c8e7 d0e0
3k5/9/9/9/3nR1p1C/9/9/9/9/4K4 w moves e5e9 d9d8 e9e5 d8d9

**Comment by HGMuller** [comment]
The first three are what comes out when you open the can of worms I mentioned before.
The first position involves subversion of the Advisor protector of the Horse against capture by the Cannon, either by pinning it, or by combining the capture with a discovered check.
The second has a plain attack on the Horse, plus a subversion of the protection of that same Horse by its King. In AXF rules it is never an issue whether a piece is protected enough; a single protector makes it free game, even when there are multiple attackers. But with a King as protector a second attacker would make the recapture illegal (rather than ineffective), and thus subverts the protection. Had there would be a Rook instead of a King on e9 moving the Rook to b9 as a second (x-ray) attacker of the Horse would be an 'idle'.
The third position involves a chase of the Cannon by subverting its protection by pinning the protector, plus a chase by activating the attacker by unpinning it. This is the one that I am least sure how to handle, because unpinning an attacker seems very similar to 'activating' the existing pseudo-legal attack by resolving pre-existing check: in both cases the attack was temporarily disabled because making the capture would leave your King in check.
The fourth perpetual seems a draw to me. The Horse is alternately exposed to an attack by the Rook and the Cannon. (I think that the fact that the Rook checks is irrelevant; moving it to e7 instead of e9 should lead to the same verdict.) But I think the crux here is that black makes no attempt to rescue the Horse in any way after the Rook attacks it. That also makes it a bit unrealistic: normally a piece would be captured on the next move when no attempt is made to solve the threat. I suppose there could be tactical reasons that tie the Rook to the e-file, so that RxH, although legal, is no good. (E.g. another red rook on e2, and a black one on a2, and something between the Horse and King to prevent the capture is accompanied by check, such as an Advisor on d7.) Perhaps the description of chasing that I gave is not complete, and should be accompanied by the rule that a new attack that is not resolved immediately by the opponent is never a chase.
essage ID:

**Comment by xiangqishare** [comment]
> 

the fourth position, N is alternately exposed to an attack by the R and the C,Most CXF refereewill think it's Perpetual chase(AXF should be written in imitation of CXF,CXF has more definitions very bad and is infinitely more complicated than AXF.,CXF referee treats it as legal study.)

**Comment by xiangqishare** [comment]
> The first three are what comes out when you open the can of worms I mentioned before. The first position involves subversion of the Advisor protector of the Horse against capture by the Cannon, either by pinning it, or by combining the capture with a discovered check. The second has a plain attack on the Horse, plus a subversion of the protection of that same Horse by its King. In AXF rules it is never an issue whether a piece is protected enough; a single protector makes it free game, even when there are multiple attackers. But with a King as protector a second attacker would make the recapture illegal (rather than ineffective), and thus subverts the protection. Had there would be a Rook instead of a King on e9 moving the Rook to b9 as a second (x-ray) attacker of the Horse would be an 'idle'. The third position involves a chase of the Cannon by subverting its protection by pinning the protector, plus a chase by activating the attacker by unpinning it. This is the one that I am least sure how to handle, because unpinning an attacker seems very similar to 'activating' the existing pseudo-legal attack by resolving pre-existing check: in both cases the attack was temporarily disabled because making the capture would leave your King in check. The fourth perpetual seems a draw to me. The Horse is alternately exposed to an attack by the Rook and the Cannon. (I think that the fact that the Rook checks is irrelevant; moving it to e7 instead of e9 should lead to the same verdict.) But I think the crux here is that black makes no attempt to rescue the Horse in any way after the Rook attacks it. That also makes it a bit unrealistic: normally a piece would be captured on the next move when no attempt is made to solve the threat. I suppose there could be tactical reasons that tie the Rook to the e-file, so that RxH, although legal, is no good. (E.g. another red rook on e2, and a black one on a2, and something between the Horse and King to prevent the capture is accompanied by check, such as an Advisor on d7.) Perhaps the description of chasing that I gave is not complete, and should be accompanied by the rule that a new attack that is not resolved immediately by the opponent is never a chase. essage ID:

I corrected mine.I mentioned
"engine and AXF disagree on chase rules.
red violation on diagram 38~41 in AFX,but Cy bugchess chessdb show draw,because both sides chase perpetually"
Actually  Bugchess→red violation,chessdb=Cy→draw

**Comment by samboy** [comment]
The intent of the Chinese Chess chasing rules is to not allow a repetition. While in “normal” international Chess a repetition is a draw, making it a draw in Chinese Chess would make that game too drawish, so they have implemented a version of the “Ko” rule in Chinese chess, to forbid repetition of position.

Of course, while it’s trivial for a computer to implement “Ko” (e.g. Zillions’ implementation of Chinese Chess makes the third repetition of a given position a lost for the player who repeats a position too many times), doing so in a human game is non-trivial, so instead we have a very subjective series of rules which amount to the same thing. Since the rules are designed for *human* and not computer play, and since, as I understand it, two different referees can come up with different judgments for a given position, it just isn’t going to be possible to get these rules exactly right in a computer implementation of Chinese Chess. 

This we can get 100% right because the rules are subjective, not objective. I personally would allow various ways to implement these rules, including a simple “Ko” implementation (not allow one to make a move that results in a position already seen).

**Comment by HGMuller** [comment]
Declaring a loss for the first player that repeats (too many times) completely changes the spirit of the rule. Which is that the player who forces the repetition should be punished. It would dramatically change the character of the game, shifting focus to gaining material by forcing the opponent to break off being chased, and surrender the chased material. Chasing plays a significant rule in xiangqi; when I played an engine that implemented the full chasing rules against one that scored all chases as a draw, 18% of the games were forfeited by chasing.
A more sensible approximate algorithm would be to determine who has the better score on his best opportunity for leaving the loop, and declare that side to lose. But that ignores the fact that 1-chases-many (or on check, one chase) is considered a valid way to force a draw. But at least it would not alter the character of the game much.
This is all a bit off topic; the issue here is how to implement the rules for the game as official game organizations decided they should be. Not to specify better or more convenient rules.
Op ma., mrt. 28, 2022 om 01:23, Sam Trenholme  schreef:
 The intent of the Chinese Chess chasing rules is to not allow a repetition. While in “normal” international Chess a repetition is a draw, making it a draw in Chinese Chess would make that game too drawish, so they have implemented a version of the “Ko” rule in Chinese chess, to forbid repetition of position.
 Of course, while it’s trivial for a computer to implement “Ko” (e.g. Zillions’ implementation of Chinese Chess makes the third repetition of a given position a lost for the player who repeats a position too many times), doing so in a human game is non-trivial, so instead we have a very subjective series of rules which amount to the same thing. Since the rules are designed for human and not computer play, and since, as I understand it, two different referees can come up with different judgments for a given position, it just isn’t going to be possible to get these rules exactly right in a computer implementation of Chinese Chess.
 This we can get 100% right because the rules are subjective, not objective. I personally would allow various ways to implement these rules, including a simple “Ko” implementation (not allow one to make a move that results in a position already seen).

**Comment by samboy** [comment]
@H.G. Huller Thank you for the reply.  

I have looked at a famous Chess game to show if implementing Ko would stop a piece from being needlessly chased (a perpetual check); I’m using Chess so the example is more clear since it’s a more well known game.

In this 1960 game played at Leipzig, Fischer was White and Tal was Black:

1. e4 e6 
2. d4 d5 
3. Nc3 Bb4 
4. e5 c5 
5. a3 Ba5 
6. b4 cxd4 
7. Qg4 Ne7 
8. bxa5 dxc3 
9. Qxg7 Rg8 
10. Qxh7 Nbc6 
11. Nf3 Qc7
12. Bb5 Bd7 
13. O-O O-O-O 
14. Bg5 Nxe5 
15. Nxe5 Bxb5 
16. Nxf7 Bxf1 
17. Nxd8 Rxg5 
18. Nxe6 Rxg2+ 
19. Kh1 Qe5 
20. Rxf1 Qxe6
21. Kxg2

And we now have this position with Black to play:

-+ are empty squares, upper case is Black, lower case is White

```
-+K+-+-+
PP+-N-+q
-+-+Q+-+
p-+P+-+-
-+-+-+-+
p-P-+-+-
-+p+-pkp
+-+-+r+-
```

For people who want to see it on a graphic chessboard, check out https://www.chessgames.com/perl/chessgame?gid=1008399 or https://lichess.org/AOw7cUIn#41

Now, let’s start the chase; White (Fischer) is ahead, but Black (Tal) can force a draw by chasing White’s king:

21... Qg4+ (Draw was declared in the played game at this point, but let’s play it out)  The Black Queen is on g4, and the White King is on g2; I will note that as Q@g4 K@g2 to keep things compact as we continue this perpetual check
23. Kh1 (Q@g4 K@h1) Qf3+ (Q@f3 K@h1)
24. Kg1 (Q@f3 K@g1) Qg4+ (Q@g4 K@g1)
25. Kh1 (Q@g4 K@h1 -- a repetition of the position in move 22, so not allowed with a simple Ko rule) Qf3+
26. Kg1 Qg4+ (and it’s now an official draw, after the third repetition)

OK, so I think we can agree that Black started the chase to draw a lost position, and that White was being chased. That in mind, let us imagine a Chess variant with a “Ko” rule.

Tal looks at the position and plays Qg4+. Fisher plays the only legal move, Kh1.
Tal now plays Qf3+ Fisher has to play Kg1 Tal then plays Qg4+ -- a position not seen yet. Now, Fischer is forced to, in this chase, repeat a previous position, but it’s Tal, *not* Fischer who started the chase.

In Chess variant with Chinese chess’s chasing rules, since it’s Tal who started the chase, Tal needs to make a different move to stop the repetition of moves, even though it’s Fischer who first repeats a position seen before.

OK, so we agree Tal should be the one to stop the chase; that’s easy for a human to see. Now, how to make objective rules so a computer sees it? That’s why this discussion is so long.

**Comment by HGMuller** [comment]
Your example is not really relevant, because the discussion is about chasing, not about checking. Bot in any case you can see that the simple Ko rule does not work: it is white that repeats first here, because his King was on g2 during the first check, rather than g1. While it is obvious that black is the player that forces the repeat. Not only to a human, but also to a computer: on every move of the repeat loop white is in check.
The point is that determining whether a player is in check or not is an absolutely objective task, as easily performed by a human as by a computer. And that the rules of the game guarantee the attack on the King can never have been pre-existing, and that it does not matter whether the checking piece is pinned to its own King, or whether the King is protected. So adjudicating perpetual checks is always trivial.
But to outlaw perpetually renewing attacks on any other piece you somehow have to define what a chase is: unlike a check, the concept of a chase is not defined in the basic (FIDE-like) rules. Not every pseudo-legal attack on a non-royal piece can be considered a threat that has to be acted on in order to prevent significant loss of material. (It is even subjective what 'significant' means here; apparently loss of an unpassed Pawn was not considered significant by those who drew up the rules.) The attack might not be legal, the attack can be futile because the attacked piece is protected and of equal or lower value than the attacker, the attack might have been pre-existing. To judge this requires a judgement on piece values, which also have no basis in the game rules.

**Comment by samboy** [comment]
Checking *is* a form a chasing. But, yes, this has become a side show and not relevant to the Xiangqi situation.

**Comment by ziyue123456789** [comment]
作者你好，对于Fairy-Stockfish-nnue存在的长久追逐规则问题，我查找了一些资料，在附件中有我上传的两个文档，这两个文档来源于2020年中国象棋竞赛规则电子版，它们包含长久追逐的一些术语解释，以及一些象棋追逐局面的具体示例，希望能对弥补Fairy-Stockfish-nnue的规则缺陷起到帮助！
[2020中国象棋追逐规则.pdf](https://github.com/ianfab/Fairy-Stockfish/files/8418924/2020.pdf)
[象棋追逐规则具体实例.pdf](https://github.com/ianfab/Fairy-Stockfish/files/8418928/default.pdf)

**Comment by cloudfish** [comment]
> 作者你好，对于Fairy-Stockfish-nnue存在的长久追逐规则问题，我查找了一些资料，在附件中有我上传的两个文档，这两个文档来源于2020年中国象棋竞赛规则电子版，它们包含长久追逐的一些术语解释，以及一些象棋追逐局面的具体示例，希望能对弥补Fairy-Stockfish-nnue的规则缺陷起到帮助！ [2020中国象棋追逐规则.pdf](https://github.com/ianfab/Fairy-Stockfish/files/8418924/2020.pdf) [象棋追逐规则具体实例.pdf](https://github.com/ianfab/Fairy-Stockfish/files/8418928/default.pdf)

Both documents are about CXA rules, not AXF rules. CXA rules are much harder to implement and afaik there are no mainstream Xiangqi engine that obey the CXA rules, and the CXA rules are rarely used outside of China. So the AXF rules should perhaps take precedence here.
这两个文档都是中国规则，不是亚洲规则。中国规则非常难以编程实现，而且目前应该没有象棋引擎支持中国规则（如果你知道支持中国规则的棋软，欢迎补充信息），而且国外也基本不用中国规则，所以应该是先实现亚洲规则比较合适。

**Comment by ziyue123456789** [comment]
&nbsp; &nbsp; &nbsp; 原来如此，我刚才找了关于亚洲规则的简单介绍（http://www.dpxq.com/hldcg/AXFRules/）。
&nbsp; &nbsp; &nbsp; 另外一方面就是，如果你的Fairy-Stockfish-NNUE缺少训练量的话，象棋程序的水平提高会比较缓慢，我建议可以弄一个分布式训练程序，让大家能利用自己的电脑一起参与训练，这样Fairy-Stockfish-NNUE的水平会飞速进步，有一个叫ggzero的象棋程序使用过这种方法（http://www.ggzero.cn/forum.php），当然，因为ggzero的作者并没有投入很多精力去优化程序的算法，所以即使有很多人参与跑谱训练，ggzero的水平也仅仅是免费最强，难以抗衡商业象棋软件。





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ianfab/Fairy-Stockfish"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年4月6日(星期三) 凌晨0:06
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ianfab/Fairy-Stockfish] Xiangqi chasing rules (#55)





  
作者你好，对于Fairy-Stockfish-nnue存在的长久追逐规则问题，我查找了一些资料，在附件中有我上传的两个文档，这两个文档来源于2020年中国象棋竞赛规则电子版，它们包含长久追逐的一些术语解释，以及一些象棋追逐局面的具体示例，希望能对弥补Fairy-Stockfish-nnue的规则缺陷起到帮助！ 2020中国象棋追逐规则.pdf 象棋追逐规则具体实例.pdf
  
Both documents are about CXA rules, not AXF rules. CXA rules are much harder to implement and afaik there are no mainstream Xiangqi engine that obey the CXA rules, and the CXA rules are rarely used outside of China. So the AXF rules should perhaps take precedence here.
 这两个文档都是中国规则，不是亚洲规则。中国规则非常难以编程实现，而且目前应该没有象棋引擎支持中国规则（如果你知道支持中国规则的棋软，欢迎补充信息），而且国外也基本不用中国规则，所以应该是先实现亚洲规则比较合适。

**Comment by xiangqishare** [comment]
> 

亚规都已经讨论过了，存在模糊分歧的一些概念，似乎没有什么统一意见。而国规太主观不可能编程的。
我以前测试过，旋风小虫猫的棋规都不一样，在一些特定循环下。

**Comment by ziyue123456789** [comment]
作者你好，中国象棋软件圈子里有人测试了Fairy-Stockfish-NNUE与一些其他象棋引擎的对局，在224次棋局中，有28次出现了由于Fairy-Stockfish-NNUE长久追逐导致输棋的情况，我想把这些长久追逐的棋谱提交给你，棋谱我放在这个链接中，（https://wwu.lanzoul.com/irhAs031d2di），再次谢谢你为开源象棋程序做出的贡献！

**Comment by ianfab** [comment]
Thanks. Could someone please give a brief summary of which are the most frequently occurring cases of chasing, i.e., which ones are most important to implement?

**Comment by HGMuller** [comment]
I never took statistics of that. Virtually all chasing is through 4-ply loops, though. Pretty common is chasing of H or C by R, presumably because R is the most powerful piece, H is pretty clumsy, and C has the misfortune of moving the same as R, which limits the choice for getting out of its path. Chases by C through moving the mounts are also pretty common.
I am not sure whether ther will be any solace in a partial implementation. Because even to recognize the most common cases you still have to do all the tests. Chasing of R by C or H is relatively easy, because it doesn't matter whether the R is protected. But you would still have to ascertain the attacker is not pinned. (And it is more difficult for a H to chase a Rook than vice versa, so presumably it will occur not so frequently.) For attacks on other pieces you would still have to test whether the piece is protected, and whether that recapture would be legal rather than just pseudo-legal.
It is presumably not very common that pieces are pinned, so a simplification would be to forego the legality tests, and count all pseudo-legal attackers and protectors.

**Comment by PraseodymiumSpike** [comment]
You’re probably going to think I’m having a high fever, but… I think
chasing should be left to third-party human adjudication. The rules are
simply too nuanced for a computer.

**Comment by cloudfish** [comment]
> Thanks. Could someone please give a brief summary of which are the most frequently occurring cases of chasing, i.e., which ones are most important to implement?

I have summarized perpetual chasing rules supported by eleeye in https://github.com/ianfab/Fairy-Stockfish/issues/55#issuecomment-953972294 . These will cover most cases.
Although the infrastructures of fairy-sf and eleeye are completely different, the source code of eleeye's perpetual checking detection can be useful. It's at https://github.com/xqbase/eleeye/blob/master/eleeye/genmoves.cpp (the PositionStruct::Protected and PositionStruct::ChasedBy functions)

**Comment by xiangqishare** [comment]
> 

你好，以下是我对亚规的个人理解，不过我不太懂英语，没法以英文形式写出。
亚规中的一些没说清楚的地方，我就按可能的我认为的编写者本意精确了，可能存在疏漏谬误。
亚规图39到41我是圆不回来，黑不算捉的理由和图105矛盾了，我个人理解抛弃图39到41。
以下大写RCNKP分别为车炮马帅(将) 兵(卒)

   1.有根、无根的判断：
一方的x子如果吃了另一方的除了K和未过河P外的y子，如果另一方有其他子可以立刻合法地吃了x，则y是有根子。如果另一方没有子能立刻合法地吃了x，则y是无根子。
    2.吃子关系：
其中一方对另一方的吃子关系为：视同轮到这一方走棋(如果这一方已经造成将军，暂且无视之)，这一方任意一个除了K和P外的x子可以合法地吃掉对方的无根子y(包含无根R)和有根R，则吃子关系为x→y或x→有根R。如果有更多的x/y/有根R，则分别形成一对一的吃子关系，比如x→y，x→有根R，x2→y，x2→y2....，但是R→有根R的关系是不存在的。
同时R→无根R/N→N/C→C的吃子关系必须再判断：轮另一方走棋，另一方的无根R/N/C能否合法地吃掉这一方的R/N/C，如果能，则相关的吃子关系不成立。
(上述吃子关系的每个子都是独一无二的，并不是兵种)
  3.长将、长捉的判断：
3.1 长将：循环局面中，如果一方每步都是将军，则该方为长将。
3.2 长捉：循环局面中，把一方走子后对另一方的吃子关系和其走子前对另一方的吃子关系进行比较，是否存在新增的吃子关系。
如果该方每一步都有新增的吃子关系，则比较吃子关系中的被吃子，所有新增吃子关系中均存在至少一个相同的被吃子，则该方为长捉。
4.循环一方违规或者和棋的判断：
三个级别按严重度：长将>长捉>非长将或长捉。
双方均为相同级别的着法则和棋。(若一方同时存在长将和长捉，视为长将)

**Comment by HGMuller** [comment]
If you want to make a partial implementation that catches many cases, and don't care whether the wrong calls it makes are overlooking chases or seeing false chases, the obvious (and huge) simplification would be to omit legality checking of the moves. Most pseudo-legal moves are also legal, so the majority of the calls would be correct.
For efficiency:
* only start classifying moves as chases along the repeat loop when you reach a repetition, since repetitions are rare.
* diversify the chases per target. A check is a check, but a chase of a Rook is not the same as a chase of another Rook. There are 15 different pieces that could be chased. If you record this with 16 bit-flags packed in a word, (which could be a bitboard, if you correct for displacement of the moving pieces) you can AND all words for the positions in the loop with the same side to move to see if he was being chased. You can abort this loop whenever the word reaches 0.

* at the core of the algorithm is a routine that determines new (pseudo-legal) attacks: direct attacks by the moved piece from its new location, discovered attacks by Rooks, Cannons, Horses or Elephants over the evacuated square, and Cannon moves over the destination square. Note that sliders only have new direct attacks in the direction perpendicular to their move.

* any new pseudo-legal attack on the King is a check.
* other direct attacks by P or K don't need to be recorded, as P and K are allowed to chase. (But P is not allowed to perpetually check!) The discovered or activated attacks made by their moves should be counted, though.

* attacks on an unpassed P can be ignored. You could clear their bits in the chase-flag word to start with, to make it reach 0 as quickly as possible.
* testing for protection is expensive; better leave it for a second iteration through the repeat loop, in the hope that no pieces were newly attacked on every move of the repeat loop. Only after the first iteration leaves us with such 'chase candidates' (and the King is not one of those) we have to do iteration 2.
* If a chase candidate is a Rook, and there is a Horse or Cannon amongst the new attackers, there is no need to test for protection, and the chase stands.
* Attackers that are of equal type as the chase candidate can be ignored, except in the case H->H where the reverse attack is blocked. If there were no other attackers the chase is discarded. In an algorithm purely based on pseudo-legal moves this could already have been done in iteration 1, but in a full algorithm it would also have to be tested whether the reverse capture is legal. Which is expensive, and therefore better deferred to iteration 2, in the hope it doesn't have to be done at all.
* For the remaining attacks we have to determine whether the chase candidate was protected. This has to be done for each attack separately, as pieces can be protected against capture by one attacker, but not of the other. In particular, one of the attackers could block an X-ray protector. (In a full algorithm the capture with one of the attackers could discover a pin of the protector that would be good against other captures.) So you have to judge whether recapture is possible after the capture by the new attacker is made. In the pseudo-legal algorithm: if a protector exists in the position before any capture, the chase is discarded. If no protector exist, and there are E, or H amongst the attackers, the chase stands. In the remaining cases all A, C and R attacks have to be tested for an X-ray protector. (A can discover a H!)
* Discarded chases are masked out of the chase-flags word, and when it reaches 0 that playerw as not chased. If at the end of the loop a non-zero remains, that player was chased.
This algorithm can be easily upgraded to one based on legal moves, by inserting legality testing for all of the pseudo-legal moves. First test whether recaptures or self-defense captures are legal, as usually they are, and that would then rule out the chase without further testing. Only for defenseless pieces you would have to test whether the attack itself was legal. (Which has only a low probability to rule out the chase.)

**Comment by zyshaoaok** [comment]
[Chess rules issues.zip](https://github.com/ianfab/Fairy-Stockfish/files/8550433/Chess.rules.issues.zip)

**Comment by xiangqishare** [comment]
> &nbsp; &nbsp; &nbsp; 原来如此，我刚才找了关于亚洲规则的简单介绍（[http://www.dpxq.com/hldcg/AXFRules/）。](http://www.dpxq.com/hldcg/AXFRules/%EF%BC%89%E3%80%82) &nbsp; &nbsp; &nbsp; 另外一方面就是，如果你的Fairy-Stockfish-NNUE缺少训练量的话，象棋程序的水平提高会比较缓慢，我建议可以弄一个分布式训练程序，让大家能利用自己的电脑一起参与训练，这样Fairy-Stockfish-NNUE的水平会飞速进步，有一个叫ggzero的象棋程序使用过这种方法（[http://www.ggzero.cn/forum.php），当然，因为ggzero的作者并没有投入很多精力去优化程序的算法，所以即使有很多人参与跑谱训练，ggzero的水平也仅仅是免费最强，难以抗衡商业象棋软件。](http://www.ggzero.cn/forum.php%EF%BC%89%EF%BC%8C%E5%BD%93%E7%84%B6%EF%BC%8C%E5%9B%A0%E4%B8%BAggzero%E7%9A%84%E4%BD%9C%E8%80%85%E5%B9%B6%E6%B2%A1%E6%9C%89%E6%8A%95%E5%85%A5%E5%BE%88%E5%A4%9A%E7%B2%BE%E5%8A%9B%E5%8E%BB%E4%BC%98%E5%8C%96%E7%A8%8B%E5%BA%8F%E7%9A%84%E7%AE%97%E6%B3%95%EF%BC%8C%E6%89%80%E4%BB%A5%E5%8D%B3%E4%BD%BF%E6%9C%89%E5%BE%88%E5%A4%9A%E4%BA%BA%E5%8F%82%E4%B8%8E%E8%B7%91%E8%B0%B1%E8%AE%AD%E7%BB%83%EF%BC%8Cggzero%E7%9A%84%E6%B0%B4%E5%B9%B3%E4%B9%9F%E4%BB%85%E4%BB%85%E6%98%AF%E5%85%8D%E8%B4%B9%E6%9C%80%E5%BC%BA%EF%BC%8C%E9%9A%BE%E4%BB%A5%E6%8A%97%E8%A1%A1%E5%95%86%E4%B8%9A%E8%B1%A1%E6%A3%8B%E8%BD%AF%E4%BB%B6%E3%80%82)
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "ianfab/Fairy-Stockfish" ***@***.***&gt;; 发送时间:&nbsp;2022年4月6日(星期三) 凌晨0:06 ***@***.***&gt;; ***@***.******@***.***&gt;; 主题:&nbsp;Re: [ianfab/Fairy-Stockfish] Xiangqi chasing rules (#55) 作者你好，对于Fairy-Stockfish-nnue存在的长久追逐规则问题，我查找了一些资料，在附件中有我上传的两个文档，这两个文档来源于2020年中国象棋竞赛规则电子版，它们包含长久追逐的一些术语解释，以及一些象棋追逐局面的具体示例，希望能对弥补Fairy-Stockfish-nnue的规则缺陷起到帮助！ 2020中国象棋追逐规则.pdf 象棋追逐规则具体实例.pdf Both documents are about CXA rules, not AXF rules. CXA rules are much harder to implement and afaik there are no mainstream Xiangqi engine that obey the CXA rules, and the CXA rules are rarely used outside of China. So the AXF rules should perhaps take precedence here. 这两个文档都是中国规则，不是亚洲规则。中国规则非常难以编程实现，而且目前应该没有象棋引擎支持中国规则（如果你知道支持中国规则的棋软，欢迎补充信息），而且国外也基本不用中国规则，所以应该是先实现亚洲规则比较合适。 — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: ***@***.***&gt;



> [Chess rules issues.zip](https://github.com/ianfab/Fairy-Stockfish/files/8550433/Chess.rules.issues.zip)

看见加我qq1272220882
